Maintenance of docker-compose Configuration Files

mcdonnnj opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

This PR does two things:
1. Update the `version` of the `docker-compose` configuration files to the minimum required version.
2. Make sure all configuration pass `docker-compose --file <filename> config --quiet`.
<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and Context ##

The first point is a result of the following error message noticed in production:

```console
docker-compose up -d
ERROR: Version in "./docker-compose.yml" is unsupported. You might be seeing this error because you're using the wrong Compose file version. Either specify a supported version (e.g "2.2" or "3.3") and place your service definitions under the `services` key, or omit the `version` key and place your service definitions at the root of the file to use version 1.
For more on the Compose file format versions, see https://docs.docker.com/compose/compose-file/
```

After discussion, we are updating to use the minimum required version, which I have determined per feature introductions in [the documentation](https://docs.docker.com/compose/compose-file/compose-versioning/#version-3).

The second point is a result of attempting to head off future issues preemptively by making sure they pass a configuration validation. When running the command listed above, the files updated returned the following error:

```console
ERROR: The Compose file is invalid because:
Service mailer has neither an image nor a build context specified. At least one must be provided.
```

<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##
`pytest` passes locally, `pre-commit` has some grief in unrelated files. I will need to do a PR to merge [cisagov/skeleton-python-library](https://github.com/cisagov/skeleton-python-library) correctly so the automated testing workflows function as expected. This is documented in #80 .
<!-- How did you test your changes? How could someone else test this PR? -->

<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

## ✅ Checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

* [x] This PR has an informative and human-readable title.
* [x] Changes are limited to a single goal - _eschew scope creep!_
* [x] _All_ future TODOs are captured in issues.
* [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
* [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).

mcdonnnj committed:
Set docker-compose file version to the minimum required

We noticed an issue after redeployment where the version specified in the
docker-compose.yml file was not supported, and docker-compose would not run the
compose configuration. This was remedied in production by setting the version
to '3.3'. After discussion with @jsf9k, I am setting these compose files to the
minimum docker-compose version they require.

The
- docker-compose.bod.yml
- docker-compose.cyhy-notification.yml
- docker-compose.cyhy.yml
files use long syntax volume mounts so they require version '3.2', and
- docker-compose.yml
uses the secrets parameter, so it only requires version '3.1'.

mcdonnnj committed:
Add image specification to docker-compose configuration files

When using docker-compose config to check all of the configurations in this
repository, the three files updated had the following result:

ERROR: The Compose file is invalid because:
Service mailer has neither an image nor a build context specified. At least
one must be provided.

I have resolved this by copying the image label from docker-compose.yml to
match.

hillaryj commented on docker-compose.yml:
```diff
@@ -1,5 +1,5 @@
 ---
-version: '3.7'
+version: '3.1'
```
Why is this one 3.1 instead of 3.2?

jsf9k replied:
@hillaryj, see the commit message for 90ae685.

hillaryj replied:
Perfect, thank you! I'll check the next time.

hillaryj commented:
I added the `blocked` label since the testing workflow is borked until #80 fixes inheritance

mcdonnnj commented:
> 
> 
> I added the `blocked` label since the testing workflow is borked until #80 fixes inheritance

Resolving #80 would involve removing the files being updated here, so I would like consensus on these changes before I kill this PR instead.

jsf9k commented on docker-compose.yml:
```diff
@@ -1,5 +1,5 @@
 ---
-version: '3.7'
+version: '3.1'
```
It's not clear to me whether we can set `docker-compose.yml` to version 3.1 and the other YAML files to version 3.2.  When this code is deployed we create a `docker-compose.override.yml` symlink to one of the `docker-compose.*.yml` files, and [that content is automatically munged with the output of `docker-compose.yml`](https://docs.docker.com/compose/extends/).

When I create the symlink and run `docker-compose config` the output indicates that the whole shebang gets version 3.1, which won't work.

mcdonnnj replied:
Ah, I overlooked that, and I believe you are correct. I updated and added a comment to explain in d63ce80.

jsf9k reviewed (changes requested):
I have a question that needs to be resolved before I can approve this PR.

mcdonnnj committed:
Update docker-compose.yml version and add comment

Update to the lowest version required by all docker-compose configurations in
this project. Added a comment explaining why this change is necessary.

jsf9k reviewed (approved):
Great job! :1st_place_medal:

dav3r reviewed (approved):
Thanks for getting us squared away here! 🦅

felddy reviewed (approved):
❄️ Let it snow!

mcdonnnj commented:
I am removing the `blocked` label because of a few reasons:

1. This fix is needed to allow BOD reporting to run without manually updating the compose files in the deployment.
2. Only `docker-compose.yml` would be validated outside of YAML checks in our modern linting/testing configuration.
3. We've put a pause on some CyHy development for now, so resolving #80 is at the least low priority for now.
